### PR TITLE
list.rb: use CLI::Parser and improve handling of flags

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

`brew list`:
* When --multiple is specified without --versions it results in the
execution of the `ls` command with `--multiple` flag, which results in a
failure. Here we assume that `--versions` flag was forgotten when  a lone`--multiple` option is provided.

* Use OpenStruct to handle arguments for `list`

**Note**: alternative solution to setting mode.versions? to true (line 39) would be to `odie` instead of `opoo` on line 38.

### CODE REVIEW

**Comment by woodruffw** [comment]
> Note: alternative solution to setting mode.versions? to true (line 39) would be to odie instead of opoo on line 38.

:+1: on `odie` -- that `send` is a little confusing.

**Comment by maxim-belkin** [comment]
well, in that case it will go away! 👍

**Comment by GauthamGoli** [comment]
@maxim-belkin Please use `Homebrew::CLI::Parser` for parsing args, it uses `OpenStruct`

File: Library/Homebrew/cmd/list.rb [hunk 1]
```diff
@@ -35,18 +50,18 @@ def list
...
+      if args.full_name?
         full_names = Formula.installed.map(&:full_name).sort(&tap_and_name_comparison)
         return if full_names.empty?
         puts Formatter.columns(full_names)
       else
         ENV["CLICOLOR"] = nil
         exec "ls", *ARGV.options_only << HOMEBREW_CELLAR
       end
-    elsif ARGV.verbose? || !$stdout.tty?
+    elsif args.verbose? || !$stdout.tty?
```
Comment (GauthamGoli) [inline Library/Homebrew/cmd/list.rb]: `Homebrew.args.verbose?`

File: Library/Homebrew/cmd/list.rb [hunk 2]
```diff
@@ -35,18 +50,18 @@ def list
       return
     end
 
-    if ARGV.include?("--pinned") || ARGV.include?("--versions")
-      filtered_list
+    if args.pinned? || args.versions?
+      filtered_list(args)
```
Comment (GauthamGoli) [inline Library/Homebrew/cmd/list.rb]: Could make `args` -> `@args`, instance variable, then its not needed to pass it as an argument.

File: Library/Homebrew/cmd/list.rb [hunk 3]
```diff
@@ -20,13 +20,28 @@
 
 require "metafiles"
 require "formula"
+require "cli_parser"
 
 module Homebrew
   module_function
 
   def list
+    args = Homebrew::CLI::Parser.parse do
```
Comment (reitermarkus) [inline Library/Homebrew/cmd/list.rb]: No need for the `Homebrew::` prefix, since this is already inside it.

File: Library/Homebrew/cmd/list.rb (hunk 1 above)
Comment (maxim-belkin) [reply Library/Homebrew/cmd/list.rb]: what is the purpose of adding `Homebrew.` ?

Comment (GauthamGoli) [reply Library/Homebrew/cmd/list.rb]: Options like `verbose`, `force`, `quiet` and `debug` are project wide, hence they are stored as global vars to be accessible everywhere. 

If we have `--verbose` passed to a command and that calls method defined in some other file, it might also need `--verbose` and we don't want to be passing around them as parameters if possible.

File: Library/Homebrew/cmd/list.rb [hunk 4]
```diff
@@ -20,13 +20,28 @@
...
+    @args = CLI::Parser.parse do
+      switch "--unbrewed"
+      switch "--pinned"
+      switch "--versions"
+      switch "--full-name"
+      switch "--multiple"
+      switch :verbose
+    end
+
+    if args.multiple? && !args.versions?
```
Comment (GauthamGoli) [inline Library/Homebrew/cmd/list.rb]: You'd have to use `@args` everywhere else

File: Library/Homebrew/cmd/list.rb [hunk 5]
```diff
@@ -35,18 +50,18 @@ def list
...
+      if @args.full_name?
         full_names = Formula.installed.map(&:full_name).sort(&tap_and_name_comparison)
         return if full_names.empty?
         puts Formatter.columns(full_names)
       else
         ENV["CLICOLOR"] = nil
         exec "ls", *ARGV.options_only << HOMEBREW_CELLAR
       end
-    elsif ARGV.verbose? || !$stdout.tty?
+    elsif Homebrew.@args.verbose? || !$stdout.tty?
```
Comment (GauthamGoli) [inline Library/Homebrew/cmd/list.rb]: This will still be `Homebrew.args.verbose?`

**Comment by GauthamGoli** [comment]
Looks good. Can you please `squash` the commits into a single/couple commit(s) ?

**Comment by maxim-belkin** [comment]
Is `opoo` message fine? Or should I change it to `odie` (and remove the line after it)?

**Comment by woodruffw** [comment]
I'm still :+1: on making it `odie`, but will defer to @GauthamGoli on it :slightly_smiling_face:

**Comment by GauthamGoli** [comment]
`brew list` can take other undeclared arguments like `-t` which would be passed to `ls`, there's bigger problem here, it's because of the option parser itself. It never accepts undeclared options and passing some undeclared option causes it to output invalid option error.

I guess, until there's solution for this, we should keep this PR on hold.

**Comment by maxim-belkin** [comment]
does `cli_parser` allow something like `flag "*"`?

**Comment by maxim-belkin** [comment]
Looks like `flag "-[ABCFGHLOPRSTUWabcdefghiklmnopqrstuwx1]*"` works, no?

**Comment by GauthamGoli** [comment]
We'd prefer not to declare options of `ls` inside brew

**Comment by maxim-belkin** [comment]
makes sense 

I guess the solution is to define acceptable `LS_FLAGS` (different on Mac and Linux) and use them here.

**Comment by GauthamGoli** [comment]
Better solution is to add an optional feature to Parser to not fail when undeclared options are passed, rather store them separately.

**Comment by maxim-belkin** [comment]
1. I think making Parser not fail for undeclared options is one part of the problem, whereas limiting acceptable (undeclared) flags for every command that accepts such flags is another.

2. In the current implementation, it is not possible to pass flags that Parser knows: `-f`, `-v`, `-q`, or `-d`. How about `--lsflags` switch that would pass all the remaining args to `ls`?

**Comment by GauthamGoli** [comment]
Point `1` is solved, I have a working implementation for making Parser not fail for undeclared options.

I am yet to find a way to solve point number `2` and I think its inherent to `OptionParser` hence difficult to override.

**Comment by maxim-belkin** [comment]
I haven't seen your implementation, but make sure it doesn't let users pass `; /bin/rm -rf / 2>/dev/null` as an option to `ls` :)

**Comment by GauthamGoli** [comment]
User can always specify `; rm -rf /` the shell itself would interpret it as a command and we wouldn't have control over it. So its not command injection. Also, as we are not offering `brew as a Service` on a server, we wouldn't have to worry about command injection :)

**Comment by GauthamGoli** [comment]
Here’s snippet from `OptionParser`’s source
```ruby
begin
  sw, = complete(:short, opt)
  # short option matched.
  val = arg.sub(/\A-/, '')
  has_arg = true
rescue InvalidOption
  # if no short options match, try completion with long
  # options.
  sw, = complete(:long, opt)
  eq ||= !rest
end
```
The above code makes undeclared short options (Ex: `-f` ) to match with long options (Ex: `--filename` ) instead of raising `InvalidOption` , as a result leads to some issues.  Couple of issues mentioned by @maxim-belkin https://github.com/Homebrew/brew/pull/4067#issuecomment-381741609

None of the OptionParsers from other standard libraries have this behaviour (Even ruby’s `Getoptlong`, and python’s `argparse` ) 

I have also raised a discussion with `OptionParser`s maintainer here https://bugs.ruby-lang.org/issues/14692 , asking him if its possible for a workaround/feature implementation to suppress this behavior.

There a no way to suppress this default behaviour, I think. 
- I tried implementing this functionality outside `OptionParser` but there are many edge cases, makes the code hard to read/maintain. Basically it is almost like writing `OptionParser`'s logic.
- We could use `Getoptlong` instead of `OptionParser` but would lose(miss?) a lot of "off-the-shelf" functionality provided by `OptionParser` which we might want to use in future.
- Don't use `Homebrew::CLI::Parser` for `brew list` until there's a work around.

 What’s the best way forward / any suggestions anyone? 
cc @MikeMcQuaid @alyssais

**Comment by ilovezfs** [comment]
yikes.

**Comment by reitermarkus** [comment]
You could use a refinement to monkey-patch that part of `OptionParser`, but you'd have to copy-paste the whole `parse_in_order` method since the relevant part is deep inside that method.

> We could use `Getoptlong` instead of `OptionParser` but would lose(miss?) a lot of "off-the-shelf" functionality provided by `OptionParser` which we might want to use in future.

Which “off-the-shelf” functionality does `OptionParser` provide?

**Comment by MikeMcQuaid** [comment]
> Point 1 is solved, I have a working implementation for making Parser not fail for undeclared options.

This should be something that is not on by default.

> I am yet to find a way to solve point number 2 and I think its inherent to OptionParser hence difficult(edit: not possible?) to override.

I think this is fine. I don't think it's essential to be able to pass those options through to `ls` and I think `brew` handling those options consistently is better.

**Comment by GauthamGoli** [comment]
Yep, monkey patching would work, but we'd have to be mindful of changes in that library whenever we change Ruby version. 

> Which “off-the-shelf” functionality does OptionParser provide?

Data types support for arguments , Documentation/help text generation. 
We'd have to implement these ourselves, meaning more maintenance

**Comment by MikeMcQuaid** [comment]
> Yep, monkey patching would work for now, but we'd have to be mindful of changes in that library whenever we change Ruby version.

I'm a pretty hard 👎 on any Ruby internal class monkey patching. I'd sooner deprecate/disable the "pass arbritrary options to `brew ls`" than do that.

**Comment by reitermarkus** [comment]
> I'd sooner deprecate/disable the "pass arbitrary options to `brew ls`" than do that.

I for one have never used this feature.

**Comment by maxim-belkin** [comment]
🍿 &nbsp;  🤔 &nbsp;  😕   &nbsp;   :trollface: 
----
 I'm +1 for disabling arbitrary options for `brew ls`.

**Comment by maxim-belkin** [comment]
what about `opoo` _vs._ `odie`? 🙄 

https://github.com/Homebrew/brew/pull/4067/files#diff-a1ae5d36c5006e319d3b6d3c8d07f58fR39

**Comment by reitermarkus** [comment]
> what about `opoo` _vs._ `odie`? 🙄 

Might want to wait for https://github.com/Homebrew/brew/pull/3987, which allows specifying conflicting and depending options.

**Comment by MikeMcQuaid** [comment]
@GauthamGoli Any news on this?

**Comment by GauthamGoli** [comment]
@MikeMcQuaid  My bad. I'll get back on this today evening.

File: Library/Homebrew/cmd/list.rb [hunk 6]
```diff
@@ -20,13 +20,28 @@
...
+    @args = CLI::Parser.parse do
+      switch "--unbrewed"
+      switch "--pinned"
+      switch "--versions"
+      switch "--full-name"
+      switch "--multiple"
+      switch :verbose
+    end
+
+    if @args.multiple? && !@args.versions?
```
Comment (GauthamGoli) [inline Library/Homebrew/cmd/list.rb]: This if block can be removed. 
Specifying constraints on options is now possible. You can have a look at `cli_parser_spec.rb`  to look at corresponding test cases.

File: Library/Homebrew/cmd/list.rb [hunk 7]
```diff
@@ -20,13 +20,28 @@
...
 module Homebrew
   module_function
 
   def list
+    @args = CLI::Parser.parse do
+      switch "--unbrewed"
+      switch "--pinned"
+      switch "--versions"
+      switch "--full-name"
+      switch "--multiple"
```
Comment (GauthamGoli) [inline Library/Homebrew/cmd/list.rb]: `switch "--multiple", required_for: "--versions"`

File: Library/Homebrew/cmd/list.rb (hunk 6 above)
Comment (maxim-belkin) [reply Library/Homebrew/cmd/list.rb]: do I need to add anything here? If so, what is the difference between `required_for` and `depends_on` (except for expressing opposite direction of a relationship) ?

File: Library/Homebrew/cmd/list.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/list.rb]: Don't you mean `depends_on`? `--multiple` is optional.

File: Library/Homebrew/cmd/list.rb [hunk 8]
```diff
@@ -20,13 +20,23 @@
...
 module Homebrew
   module_function
 
   def list
+    @args = CLI::Parser.parse do
+      switch "--unbrewed"
+      switch "--pinned"
+      switch "--versions"
+      switch "--full-name"
+      switch "--multiple", required_for: "--versions"
```
Comment (reitermarkus) [inline Library/Homebrew/cmd/list.rb]: Should be `depends_on`.

Comment (maxim-belkin) [reply Library/Homebrew/cmd/list.rb]: So, what does the phrase
```
`--multiple` requires `--versions`
```
in plain English corresponds to in this new syntax?
```
switch "--multiple", depends_on: "--versions"
```
?

Comment (reitermarkus) [reply Library/Homebrew/cmd/list.rb]:
```ruby
switch "--multiple", depends_on: "--versions"
```

`--multiple` is optional and can only be given as a “sub-option” for `--versions`.


```ruby
switch "--multiple", required_for: "--versions"
```

When `--versions` is given, `--multiple` also has to be specified.

File: Library/Homebrew/cmd/list.rb (hunk 7 above)
Comment (GauthamGoli) [reply Library/Homebrew/cmd/list.rb]: Ops.. I got that wrong, you're right.

**Comment by GauthamGoli** [comment]
@maxim-belkin Sorry. I added constraints for flags but not switches. I'll get that done very soon, and then we could proceed with this.

**Comment by GauthamGoli** [comment]
@maxim-belkin Please rebase with master

**Comment by GauthamGoli** [comment]
I have tested this locally. Will be merged when @MikeMcQuaid gives 👍

**Comment by MikeMcQuaid** [comment]
Thanks @maxim-belkin! If people go 😭 over the death of custom flags we'll think about fiddling with this more but I'm not sure it's worth blocking this PR over.

**Comment by MikeMcQuaid** [comment]
> -1 is not necessary when piping output to another command.

Regardless, given it's used by others, we should fix this.
